src/optics/focusAtom.ts

```diff
@@ -0,0 +1,74 @@
+import { atom } from 'jotai'
+import * as O from 'optics-ts'
+import { PrimitiveAtom, SetStateAction, WritableAtom } from '../core/types'
+
+export type FocusAtom = {
+  <S, A>(
+    atom: PrimitiveAtom<S>,
+    callback: (optic: O.OpticFor<S>) => O.Prism<S, any, A>
+  ): WritableAtom<A | undefined, SetStateAction<A>>
+
+  <S, A>(
+    atom: PrimitiveAtom<S>,
+    callback: (optic: O.OpticFor<S>) => O.Traversal<S, any, A>
+  ): WritableAtom<Array<A>, SetStateAction<A>>
+
+  <S, A>(
+    atom: PrimitiveAtom<S>,
+    callback: (
+      optic: O.OpticFor<S>
+    ) => O.Lens<S, any, A> | O.Equivalence<S, any, A> | O.Iso<S, any, A>
+  ): PrimitiveAtom<A>
+}
+
+export const focusAtom: FocusAtom = <S, A>(
+  baseAtom: PrimitiveAtom<S>,
+  callback: (
+    optic: O.OpticFor<S>
+  ) =>
+    | O.Lens<S, any, A>
+    | O.Equivalence<S, any, A>
+    | O.Iso<S, any, A>
+    | O.Prism<S, any, A>
+    | O.Traversal<S, any, A>
+): any => {
+  const focus = callback(O.optic<S>())
+  return atom<A, SetStateAction<A>>(
+    (atomGetter) => {
+      const newValue = getValueUsingOptic(focus, atomGetter(baseAtom))
+      return newValue
+    },
+    (_, set, update) => {
+      const newValueProducer =
+        update instanceof Function
+          ? O.modify(focus)(update)
+          : O.set(focus)(update)
+
+      set(baseAtom, (oldBaseValue) => {
+        const newBaseValue = newValueProducer(oldBaseValue)
+        return newBaseValue
+      })
```

Comment:
```suggestion
      set(baseAtom, newValueProducer)
```
Can this be simplified like this assuming it's not using `this`?